### PR TITLE
Adding abillity to deploy cert to Palo Alto Networks Firewall via API.

### PR DESCRIPTION
**Pull request by paulmnguyen**

### CODE REVIEW

**Comment by paulmnguyen** [comment]
There is an error in Travis regarding "git diff --exit-code"

How do I fix this?

**Comment by Neilpang** [comment]
please add your usage to this page:
https://github.com/Neilpang/acme.sh/wiki/deployhooks

**Comment by Neilpang** [comment]
please read this guide and create a bug:
https://github.com/Neilpang/acme.sh/wiki/DNS-API-Dev-Guide#10-please-create-a-new-issue-for-future-bugs

**Comment by paulmnguyen** [comment]
I have added the usage and created a bug for future bug reports.

File: deploy/panos.sh [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+# This function is to deploy the cert to the Firewall
+deploy_cert() {
+  #Checking for IP and Key
+  if [ -z "$_panos_host" ] && [ -z "$_panos_key" ]; then
+    _err "No IP or Key found. Can not deploy certs."
+    return 1
+  else # IP and key found lets import cert, key and commit.
+    _debug "Deploying Cert"
+    cert_response=$(curl -k --form file=@"$_cfullchain" "https://$_panos_host/api/?type=import&category=certificate&certificate-name=$_cdomain&format=pem&key=$_panos_key")
```
Comment (Neilpang) [inline deploy/panos.sh]: don't use `curl`,  please use `_get()` or `_post()` function instead.
we need to support wget.

Comment (paulmnguyen) [reply deploy/panos.sh]: @Neilpang How do I make a form-data file post using your  _post() function?

Comment (Neilpang) [reply deploy/panos.sh]: it's submitting a form to upload cert and key.  we can use `_post()` to simulate this action.
It's just about the content-type and the postbody.

content-type is just a header, which is easy to set.
postbody is urlencoded file,  which is  percentage sign`%` splited hex encoding bytes.

We have a `_url_encode()` function to do so.

**Comment by paulmnguyen** [comment]
Reopening

**Comment by Neilpang** [comment]
please fix the ci errors.

File: deploy/panos.sh (hunk 1 above)
Comment (paulmnguyen) [reply deploy/panos.sh]: Thanks Neil refactored the code to use _post

File: deploy/panos.sh [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      _err "PANOS_HOST not defined."
+      return 1
+    fi
+  else
+    _debug "PANOS HOST is set. Save to domain conf."
+    _panos_host="$PANOS_HOST"
+    _savedomainconf _panos_host "$_panos_host"
+  fi
+  # Retrieve stored variables
+  _panos_user="$(_readaccountconf_mutable PANOS_USER)"
```
Comment (Neilpang) [inline deploy/panos.sh]: do not use `_readaccountconf_mutable/_saveaccountconf_mutable`,
please use `_getdeployconf/_savedeployconf` instead.

example:
https://github.com/acmesh-official/acme.sh/blob/master/deploy/docker.sh#L67

Comment (paulmnguyen) [reply deploy/panos.sh]: Should I also replace _savedomainconf as well?

Comment (paulmnguyen) [reply deploy/panos.sh]: decided to just use savedeployconf for all 3.

**Comment by paulmnguyen** [comment]
A user stated that there is an error when deploying the certificates to Panorama. I will need to resolve that issue before we can merge this code.

**Comment by paulmnguyen** [comment]
Fixed the deployment issue to panorama.

**Pull request merged**